### PR TITLE
[Perf] remove redundant sp alltoall communication for z-image

### PR DESCRIPTION
**Pull request by JustQJ**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

remove redundant alltoall communication in attention when enabling sp for z-image

## Test Plan
tesed on ascend 910B
```
python text_to_image.py --model Tongyi-MAI/Z-Image-Turbo --num_inference_steps 50 --enforce_eager --ulysses_degree 2 --height 1024 --width 1024
```

## Test Result

|image size |before time(s)| after time(s)|
|---|---|---|
|1024*1024|  13.30|12.80|
|2048*2048|  65.95|61.86|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/z_image/z_image_transformer.py [hunk 1]
```diff
@@ -291,6 +292,7 @@ def __init__(
             softmax_scale=1.0 / (self.head_dim**0.5),
             causal=False,
             num_kv_heads=self.to_qkv.num_kv_heads,
+            disable_sp=disable_sp,
```
Comment (ZJY0516) [inline vllm_omni/diffusion/models/z_image/z_image_transformer.py:295]: why not just set `disable_tp = True` here?

File: vllm_omni/diffusion/attention/parallel/factory.py [hunk 2]
```diff
@@ -15,6 +15,7 @@ def build_parallel_attention_strategy(
     scatter_idx: int,
     gather_idx: int,
     use_sync: bool,
+    disable_sp: bool = False,
```
Comment (ZJY0516) [inline vllm_omni/diffusion/attention/parallel/factory.py]: I'm a little confusing here. What does `disable_sp` mean when we use sp?

File: vllm_omni/diffusion/models/z_image/z_image_transformer.py (hunk 1 above)
Comment (ZJY0516) [reply vllm_omni/diffusion/models/z_image/z_image_transformer.py:295]: I see. We have 3 kinds of layers in z-image

**Comment by ZJY0516** [comment]
cc @mxuax

File: vllm_omni/diffusion/attention/parallel/factory.py (hunk 2 above)
Comment (JustQJ) [reply vllm_omni/diffusion/attention/parallel/factory.py]: I will remove this later

File: vllm_omni/diffusion/attention/layer.py
Comment (mxuax) [inline vllm_omni/diffusion/attention/layer.py:1]: Thx for the contribution, but could you explain which LOCs trigger the redundant all-to-all communication? In the current implementation, in build_parallel_attention_strategy, we will directly use NoParallelAttention if ring==1

Comment (mxuax) [reply vllm_omni/diffusion/attention/layer.py:1]: BTW, I see you use --ulysses_degree 2  in your test command, but sp includes both ulysses and ring, if ulysses>1 and ring ==1,  it will be handled by the following lines in build_parallel_attention_strategy
```
# Ulysses (or Hybrid Ulysses+Ring)
    if ulysses_degree > 1:
        return UlyssesParallelAttention(
            sp_group=sp_group,
            scatter_idx=scatter_idx,
            gather_idx=gather_idx,
            use_sync=use_sync,
        )
```
If  --ulysses_degree 2, again, since sp includes both ulysses and ring, use this disable_sp flag seems confusing to me :P

Comment (JustQJ) [reply vllm_omni/diffusion/attention/layer.py:1]: In z-image, if SP is enabled (i.e., setting ring or usp to 2), all layers (self.noise_refiner, self.context_refiner, and self.layers) use attention with the SP strategy in build_parallel_attention_strategy (returning UlyssesParallelAttention or RingParallelAttention), which cause communication for each layer. However, in z-image, the sp_plan is only configured for self.layers. Therefore, there is no need to enable the SP strategy for self.noise_refiner and self.context_refiner, and it should be disabled for them.

```
ZImageTransformer2DModel:

   _sp_plan = {
          # Shard unified_prepare outputs (similar to Wan's rope module)
          # This shards all 4 return values: unified, unified_cos, unified_sin, unified_attn_mask
          "unified_prepare": {
              0: SequenceParallelInput(split_dim=1, expected_dims=3, split_output=True),  # unified
              1: SequenceParallelInput(split_dim=1, expected_dims=3, split_output=True),  # unified_cos
              2: SequenceParallelInput(split_dim=1, expected_dims=3, split_output=True),  # unified_sin
              3: SequenceParallelInput(split_dim=1, expected_dims=2, split_output=True),  # unified_attn_mask
          },
          # Gather output at final_layer (default: patch_size=2, f_patch_size=1)
          "all_final_layer.2-1": SequenceParallelOutput(gather_dim=1, expected_dims=3),
      }
       self.noise_refiner = nn.ModuleList(
            [
                ZImageTransformerBlock(
                    1000 + layer_id,
                    dim,
                    n_heads,
                    n_kv_heads,
                    norm_eps,
                    qk_norm,
                    modulation=True,
                )
                for layer_id in range(n_refiner_layers)
            ]
        )
        self.context_refiner = nn.ModuleList(
            [
                ZImageTransformerBlock(
                    layer_id,
                    dim,
                    n_heads,
                    n_kv_heads,
                    norm_eps,
                    qk_norm,
                    modulation=False,
                )
                for layer_id in range(n_refiner_layers)
            ]
        )

        self.layers = nn.ModuleList(
            [
                ZImageTransformerBlock(layer_id, dim, n_heads, n_kv_heads, norm_eps, qk_norm)
                for layer_id in range(n_layers)
            ]
        )
```

Comment (mxuax) [reply vllm_omni/diffusion/attention/layer.py:1]: Okay, I see what you are doing here. This PR LGTM, you can add this to the doc /docs/user_guide/diffusion_acceleration.md to emphasize that we now support a more fine-grained sp control.

Comment (mxuax) [reply vllm_omni/diffusion/attention/layer.py:1]: Aside from the doc updation, I think this PR is ready @ZJY0516

Comment (JustQJ) [reply vllm_omni/diffusion/attention/layer.py:1]: Hi, I think there's no need to add this to the doc since there's no user-facing configuration for it.

Comment (ZJY0516) [reply vllm_omni/diffusion/attention/layer.py:1]: > Hi, I think there's no need to add this to the doc since there's no user-facing configuration for it.

It's added for developers

**Comment by ZJY0516** [comment]
I think it's not clean enough. Could we only apply `sp_plan` to `slef.layers`?

**Comment by JustQJ** [comment]
> I think it's not clean enough. Could we only apply `sp_plan` to `slef.layers`?

A bit confused. Doesn't the current implementation already do exactly that?

File: vllm_omni/diffusion/attention/parallel/factory.py (hunk 2 above)
Comment (mxuax) [reply vllm_omni/diffusion/attention/parallel/factory.py]: Plz add a todo comment here. We will merge this in the sp hooks in the future since we need nonintrusive implementation.

**Comment by mxuax** [comment]
> > I think it's not clean enough. Could we only apply `sp_plan` to `slef.layers`?
> 
> A bit confused. Doesn't the current implementation already do exactly that?

Since we hope to include all these commands in the sp plan without changing the model implementation. Thus you can just add todo label here and we will refactor this in the future

**Comment by lishunyang12** [comment]
@JustQJ Hey, the SP alltoall removal shows a nice 3.7-6.2% speedup on Z-Image at different resolutions, tested on Ascend 910B. This is a small, clean change — is there anything blocking this from being merged?

**Comment by JustQJ** [comment]
> @JustQJ Hey, the SP alltoall removal shows a nice 3.7-6.2% speedup on Z-Image at different resolutions, tested on Ascend 910B. This is a small, clean change — is there anything blocking this from being merged?

please refer to https://github.com/vllm-project/vllm-omni/pull/1275
